core: region heartbeat tracer panics with "counter cannot decrease in value"

## Bug Report

### What did you do?

`pd-server` is running with `enable-heartbeat-breakdown-metrics = true` (the default). On a production cluster hosted on a cloud VM, a region heartbeat triggered a panic that crashed the whole `pd-server` process.

The trigger is an environmental one: on some virtualized hosts the OS monotonic clock (`CLOCK_MONOTONIC`, read by Go's `time.Now()` via the vDSO) can briefly move **backwards** — e.g. unsynchronized/unstable TSC across vCPUs, or after a live migration. PD has no defense against this in the heartbeat breakdown metrics path, so a tiny backward step turns into a fatal panic.

### What did you expect to see?

A transient, sub-microsecond clock anomaly should at most produce a slightly inaccurate latency metric. It must **never** crash `pd-server`.

### What did you see instead?

```
panic: counter cannot decrease in value

goroutine [running]:
github.com/prometheus/client_golang/prometheus.(*counter).Add(...)
    .../prometheus/client_golang@v1.20.5/prometheus/counter.go:128
github.com/tikv/pd/pkg/core.(*regionHeartbeatProcessTracer).OnAllStageFinished(...)
    pkg/core/metrics.go:349
github.com/tikv/pd/server/cluster.(*RaftCluster).HandleRegionHeartbeat(...)
    server/cluster/cluster_worker.go:65
github.com/tikv/pd/server.(*GrpcServer).RegionHeartbeat(...)
    server/grpc_service.go:1360
github.com/pingcap/kvproto/pkg/pdpb._PD_RegionHeartbeat_Handler(...)
...
```

### What version of PD are you using (`pd-server -V`)?

Observed on a v8.5-based build; the offending code is identical on current `master`. The breakdown-metrics tracer was introduced in #7871, so all releases since then are affected. Line numbers below refer to `master`.

---

## Root cause analysis

`OnAllStageFinished` (`pkg/core/metrics.go:346`) does:

```go
func (h *regionHeartbeatProcessTracer) OnAllStageFinished() {
	now := time.Now()
	h.OtherDuration = now.Sub(h.lastCheckTime)
	otherDurationSum.Add(h.OtherDuration.Seconds()) // metrics.go:349
	otherCount.Inc()
}
```

`prometheus.Counter.Add(v)` panics when `v < 0`. So `h.OtherDuration` is negative, i.e. `now.Sub(h.lastCheckTime) < 0`.

In a single goroutine this should be impossible: every `lastCheckTime` is assigned from `time.Now()`, which carries a monotonic reading, and `t.Sub(u)` uses the monotonic clock when both operands have it, so the result is normally `>= 0`. The tracer is also used strictly within one goroutine (verified: the async runner closures capture only `c`/`region`/`overlaps`, never the tracer; `sync.Pool` hands distinct objects to concurrent callers; `Release()` is deferred after all use). So a code-level data race is ruled out.

The only remaining source of a negative delta is a **non-monotonic OS monotonic clock**, which is a known phenomenon on virtualized hosts (unstable TSC across vCPUs, live migration). NTP/chrony does not prevent this — it disciplines `CLOCK_REALTIME` and can only slew the *rate* of `CLOCK_MONOTONIC`, never step it backwards; the backward jump originates below NTP, at the hardware/hypervisor clocksource layer.

### Why it always panics at the `"Other"` stage

A negative only appears when `|clock backstep| > measured window`. The eight `Add` sites measure windows of very different sizes:

- `PreCheck`, `RegionGuide`, `SaveCache_*`, `CollectRegionStats` each bracket real work (mutex acquisition, B-tree overlaps/insert, validation) — windows of µs to ms, which dwarf any clock jitter.
- **`Other`** is the residual gap between the last checkpoint (`OnCollectRegionStatsFinished`) and the final `time.Now()`. With `enable-heartbeat-concurrent-runner = true` (default), the trailing `SaveRegionToKV` / `SyncRegionToFollower` work is async-enqueued, so this window is only **tens of nanoseconds** — the same order as the clock jitter.

So negatives statistically concentrate almost entirely in `"Other"`, which matches the field reports (the crash recurs at this exact line). It is a selection effect, not a coincidence.

Because the panic happens in a gRPC stream handler goroutine with no recovery interceptor, it terminates the entire `pd-server` process.

## Suggested fix

Guard every duration-derived `Counter.Add` in `pkg/core/metrics.go` against negative values (clamp to 0), e.g. via a small helper:

```go
func addDurationSum(c prometheus.Counter, d time.Duration) {
	if d > 0 {
		c.Add(d.Seconds())
	}
}
```

This downgrades a transient environmental clock glitch from a process-killing panic to at most a negligibly under-counted metric. There are 8 such sites.

Operational workaround (no restart needed): `pd-ctl config set enable-heartbeat-breakdown-metrics false` switches to the noop tracer and avoids the path entirely.

## Additional bugs found in the same area

While analyzing this, three more issues turned up in the tracer code:

1. **`pkg/core/metrics.go:274` — wrong field in `OnAsyncHotStatsFinished`.** It does `asyncHotStatsDurationSum.Add(h.preCheckDuration.Seconds())` but should add `h.asyncHotStatsDuration`. The `AsyncHotStatsDuration` metric is therefore double-counting `PreCheck`'s duration. (Incidentally this is why that stage never panics — it re-adds an already-positive value.)

2. **`pkg/core/metrics.go:312` — wrong checkpoint in `OnCheckOverlapsFinished`.** It computes `now.Sub(h.lastCheckTime)` while every other `SaveCache_*` method uses `h.saveCacheStats.lastCheckTime`. It is currently harmless only because `OnSaveCacheBegin` initializes both fields to the same value and nothing mutates `h.lastCheckTime` in between — a latent bug that would break if the call order ever changed.

3. **`pkg/mcs/scheduling/server/cluster.go:784` — missing `tracer.Release()`.** The scheduling-microservice `HandleRegionHeartbeat` obtains a tracer from the pool but never calls `defer tracer.Release()`, unlike the classic path in `server/cluster/cluster_worker.go:42`. The pooled object is never returned, so `sync.Pool` keeps allocating new tracers. Not related to the panic, but a real object leak.

--- a/pkg/core/metrics.go
+++ b/pkg/core/metrics.go
@@ -19,7 +19,6 @@ import (
 	"time"
 
 	"github.com/prometheus/client_golang/prometheus"
-	"go.uber.org/zap"
 )
 
 var (
@@ -143,15 +142,6 @@ var tracerPool = &sync.Pool{
 	},
 }
 
-type saveCacheStats struct {
-	startTime              time.Time
-	lastCheckTime          time.Time
-	checkOverlapsDuration  time.Duration
-	validateRegionDuration time.Duration
-	setRegionDuration      time.Duration
-	updateSubTreeDuration  time.Duration
-}
-
 // RegionHeartbeatProcessTracer is used to trace the process of handling region heartbeat.
 type RegionHeartbeatProcessTracer interface {
 	// Begin starts the tracing.
@@ -178,8 +168,6 @@ type RegionHeartbeatProcessTracer interface {
 	OnCollectRegionStatsFinished()
 	// OnAllStageFinished will be called when all stages are finished.
 	OnAllStageFinished()
-	// LogFields returns the log fields.
-	LogFields() []zap.Field
 	// Release releases the tracer.
 	Release()
 }
@@ -227,22 +215,15 @@ func (*noopHeartbeatProcessTracer) OnCollectRegionStatsFinished() {}
 // OnAllStageFinished implements the RegionHeartbeatProcessTracer interface.
 func (*noopHeartbeatProcessTracer) OnAllStageFinished() {}
 
-// LogFields implements the RegionHeartbeatProcessTracer interface.
-func (*noopHeartbeatProcessTracer) LogFields() []zap.Field {
-	return nil
-}
-
 // Release implements the RegionHeartbeatProcessTracer interface.
 func (*noopHeartbeatProcessTracer) Release() {}
 
+// regionHeartbeatProcessTracer measures the duration of each region-heartbeat
+// processing stage and reports it to the breakdown metrics. It keeps a single
+// checkpoint: a stage's duration is the time elapsed since the previous stage
+// advanced the checkpoint.
 type regionHeartbeatProcessTracer struct {
-	startTime             time.Time
-	lastCheckTime         time.Time
-	preCheckDuration      time.Duration
-	asyncHotStatsDuration time.Duration
-	regionGuideDuration   time.Duration
-	saveCacheStats        saveCacheStats
-	OtherDuration         time.Duration
+	lastCheckTime time.Time
 }
 
 // NewHeartbeatProcessTracer returns a heartbeat process tracer.
@@ -252,116 +233,91 @@ func NewHeartbeatProcessTracer() RegionHeartbeatProcessTracer {
 
 // Begin implements the RegionHeartbeatProcessTracer interface.
 func (h *regionHeartbeatProcessTracer) Begin() {
+	h.lastCheckTime = time.Now()
+}
+
+// measure records one processing stage: it reads the clock once, computes the
+// duration since the previous checkpoint, advances the checkpoint, and updates
+// the stage's duration-sum and count counters.
+//
+// A non-positive duration is not added to the sum counter. time.Now() normally
+// carries a monotonic reading, so two consecutive reads never go backwards; but
+// on some virtualized hosts the OS monotonic clock (CLOCK_MONOTONIC) can briefly
+// regress -- e.g. an unsynchronized TSC across vCPUs, or a live migration.
+// Without this guard the negative value would reach prometheus.Counter.Add,
+// which panics with "counter cannot decrease in value" and crashes pd-server.
+// Guarding turns that rare, external clock glitch into a negligibly
+// under-counted metric instead of a fatal panic.
+func (h *regionHeartbeatProcessTracer) measure(durationSum, count prometheus.Counter) {
 	now := time.Now()
-	h.startTime = now
+	duration := now.Sub(h.lastCheckTime)
 	h.lastCheckTime = now
+	if duration > 0 {
+		durationSum.Add(duration.Seconds())
+	}
+	count.Inc()
+}
+
+// resetCheckpoint starts a fresh measurement window at the current time. It is
+// used at the SaveCache boundaries, whose enclosed sub-stages are timed on their
+// own and should not be attributed to the surrounding stages.
+func (h *regionHeartbeatProcessTracer) resetCheckpoint() {
+	h.lastCheckTime = time.Now()
 }
 
 // OnPreCheckFinished implements the RegionHeartbeatProcessTracer interface.
 func (h *regionHeartbeatProcessTracer) OnPreCheckFinished() {
-	now := time.Now()
-	h.preCheckDuration = now.Sub(h.lastCheckTime)
-	h.lastCheckTime = now
-	preCheckDurationSum.Add(h.preCheckDuration.Seconds())
-	preCheckCount.Inc()
+	h.measure(preCheckDurationSum, preCheckCount)
 }
 
 // OnAsyncHotStatsFinished implements the RegionHeartbeatProcessTracer interface.
 func (h *regionHeartbeatProcessTracer) OnAsyncHotStatsFinished() {
-	now := time.Now()
-	h.asyncHotStatsDuration = now.Sub(h.lastCheckTime)
-	h.lastCheckTime = now
-	asyncHotStatsDurationSum.Add(h.preCheckDuration.Seconds())
-	asyncHotStatsCount.Inc()
+	h.measure(asyncHotStatsDurationSum, asyncHotStatsCount)
 }
 
 // OnRegionGuideFinished implements the RegionHeartbeatProcessTracer interface.
 func (h *regionHeartbeatProcessTracer) OnRegionGuideFinished() {
-	now := time.Now()
-	h.regionGuideDuration = now.Sub(h.lastCheckTime)
-	h.lastCheckTime = now
-	regionGuideDurationSum.Add(h.regionGuideDuration.Seconds())
-	regionGuideCount.Inc()
+	h.measure(regionGuideDurationSum, regionGuideCount)
 }
 
 // OnSaveCacheBegin implements the RegionHeartbeatProcessTracer interface.
 func (h *regionHeartbeatProcessTracer) OnSaveCacheBegin() {
-	now := time.Now()
-	h.saveCacheStats.startTime = now
-	h.saveCacheStats.lastCheckTime = now
-	h.lastCheckTime = now
+	h.resetCheckpoint()
 }
 
 // OnSaveCacheFinished implements the RegionHeartbeatProcessTracer interface.
 func (h *regionHeartbeatProcessTracer) OnSaveCacheFinished() {
-	// update the outer checkpoint time
-	h.lastCheckTime = time.Now()
-}
-
-// OnCollectRegionStatsFinished implements the RegionHeartbeatProcessTracer interface.
-func (h *regionHeartbeatProcessTracer) OnCollectRegionStatsFinished() {
-	now := time.Now()
-	regionCollectDurationSum.Add(now.Sub(h.lastCheckTime).Seconds())
-	regionCollectCount.Inc()
-	h.lastCheckTime = now
+	h.resetCheckpoint()
 }
 
 // OnCheckOverlapsFinished implements the RegionHeartbeatProcessTracer interface.
 func (h *regionHeartbeatProcessTracer) OnCheckOverlapsFinished() {
-	now := time.Now()
-	h.saveCacheStats.checkOverlapsDuration = now.Sub(h.lastCheckTime)
-	h.saveCacheStats.lastCheckTime = now
-	checkOverlapsDurationSum.Add(h.saveCacheStats.checkOverlapsDuration.Seconds())
-	checkOverlapsCount.Inc()
+	h.measure(checkOverlapsDurationSum, checkOverlapsCount)
 }
 
 // OnValidateRegionFinished implements the RegionHeartbeatProcessTracer interface.
 func (h *regionHeartbeatProcessTracer) OnValidateRegionFinished() {
-	now := time.Now()
-	h.saveCacheStats.validateRegionDuration = now.Sub(h.saveCacheStats.lastCheckTime)
-	h.saveCacheStats.lastCheckTime = now
-	validateRegionDurationSum.Add(h.saveCacheStats.validateRegionDuration.Seconds())
-	validateRegionCount.Inc()
+	h.measure(validateRegionDurationSum, validateRegionCount)
 }
 
 // OnSetRegionFinished implements the RegionHeartbeatProcessTracer interface.
 func (h *regionHeartbeatProcessTracer) OnSetRegionFinished() {
-	now := time.Now()
-	h.saveCacheStats.setRegionDuration = now.Sub(h.saveCacheStats.lastCheckTime)
-	h.saveCacheStats.lastCheckTime = now
-	setRegionDurationSum.Add(h.saveCacheStats.setRegionDuration.Seconds())
-	setRegionCount.Inc()
+	h.measure(setRegionDurationSum, setRegionCount)
 }
 
 // OnUpdateSubTreeFinished implements the RegionHeartbeatProcessTracer interface.
 func (h *regionHeartbeatProcessTracer) OnUpdateSubTreeFinished() {
-	now := time.Now()
-	h.saveCacheStats.updateSubTreeDuration = now.Sub(h.saveCacheStats.lastCheckTime)
-	h.saveCacheStats.lastCheckTime = now
-	updateSubTreeDurationSum.Add(h.saveCacheStats.updateSubTreeDuration.Seconds())
-	updateSubTreeCount.Inc()
+	h.measure(updateSubTreeDurationSum, updateSubTreeCount)
+}
+
+// OnCollectRegionStatsFinished implements the RegionHeartbeatProcessTracer interface.
+func (h *regionHeartbeatProcessTracer) OnCollectRegionStatsFinished() {
+	h.measure(regionCollectDurationSum, regionCollectCount)
 }
 
 // OnAllStageFinished implements the RegionHeartbeatProcessTracer interface.
 func (h *regionHeartbeatProcessTracer) OnAllStageFinished() {
-	now := time.Now()
-	h.OtherDuration = now.Sub(h.lastCheckTime)
-	otherDurationSum.Add(h.OtherDuration.Seconds())
-	otherCount.Inc()
-}
-
-// LogFields implements the RegionHeartbeatProcessTracer interface.
-func (h *regionHeartbeatProcessTracer) LogFields() []zap.Field {
-	return []zap.Field{
-		zap.Duration("pre-check-duration", h.preCheckDuration),
-		zap.Duration("async-hot-stats-duration", h.asyncHotStatsDuration),
-		zap.Duration("region-guide-duration", h.regionGuideDuration),
-		zap.Duration("check-overlaps-duration", h.saveCacheStats.checkOverlapsDuration),
-		zap.Duration("validate-region-duration", h.saveCacheStats.validateRegionDuration),
-		zap.Duration("set-region-duration", h.saveCacheStats.setRegionDuration),
-		zap.Duration("update-sub-tree-duration", h.saveCacheStats.updateSubTreeDuration),
-		zap.Duration("other-duration", h.OtherDuration),
-	}
+	h.measure(otherDurationSum, otherCount)
 }
 
 // Release implements the RegionHeartbeatProcessTracer interface.

--- a/pkg/core/metrics_test.go
+++ b/pkg/core/metrics_test.go
@@ -1,0 +1,84 @@
+// Copyright 2024 TiKV Project Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package core
+
+import (
+	"testing"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/testutil"
+	"github.com/stretchr/testify/require"
+)
+
+// recordingStage couples a tracer stage method with the counters it feeds, so a
+// single table can drive the tests over every measuring stage uniformly.
+type recordingStage struct {
+	name   string
+	finish func(*regionHeartbeatProcessTracer)
+	sum    prometheus.Counter
+	count  prometheus.Counter
+}
+
+func recordingStages() []recordingStage {
+	return []recordingStage{
+		{"PreCheck", (*regionHeartbeatProcessTracer).OnPreCheckFinished, preCheckDurationSum, preCheckCount},
+		{"AsyncHotStats", (*regionHeartbeatProcessTracer).OnAsyncHotStatsFinished, asyncHotStatsDurationSum, asyncHotStatsCount},
+		{"RegionGuide", (*regionHeartbeatProcessTracer).OnRegionGuideFinished, regionGuideDurationSum, regionGuideCount},
+		{"CheckOverlaps", (*regionHeartbeatProcessTracer).OnCheckOverlapsFinished, checkOverlapsDurationSum, checkOverlapsCount},
+		{"ValidateRegion", (*regionHeartbeatProcessTracer).OnValidateRegionFinished, validateRegionDurationSum, validateRegionCount},
+		{"SetRegion", (*regionHeartbeatProcessTracer).OnSetRegionFinished, setRegionDurationSum, setRegionCount},
+		{"UpdateSubTree", (*regionHeartbeatProcessTracer).OnUpdateSubTreeFinished, updateSubTreeDurationSum, updateSubTreeCount},
+		{"CollectRegionStats", (*regionHeartbeatProcessTracer).OnCollectRegionStatsFinished, regionCollectDurationSum, regionCollectCount},
+		{"Other", (*regionHeartbeatProcessTracer).OnAllStageFinished, otherDurationSum, otherCount},
+	}
+}
+
+// TestRegionHeartbeatTracerToleratesNonMonotonicClock checks the tracer is
+// robust to a backwards jump of the OS monotonic clock, which can happen on
+// virtualized hosts (an unsynchronized TSC across vCPUs, or a live migration).
+// For every stage, a checkpoint left in the future must not make the stage panic
+// or push a negative value into its duration counter (prometheus.Counter.Add
+// panics on a negative value).
+func TestRegionHeartbeatTracerToleratesNonMonotonicClock(t *testing.T) {
+	re := require.New(t)
+	for _, s := range recordingStages() {
+		h := &regionHeartbeatProcessTracer{}
+		h.lastCheckTime = time.Now().Add(time.Hour) // the clock appears to run backwards
+		before := testutil.ToFloat64(s.sum)
+		re.NotPanics(func() { s.finish(h) }, s.name)
+		re.GreaterOrEqual(testutil.ToFloat64(s.sum), before, s.name) // the counter never decreases
+	}
+}
+
+// TestRegionHeartbeatTracerRecordsEachStageIndependently checks that every stage
+// reports the time spent in that stage -- and only that stage -- into its own
+// counters, exactly once. Each stage is given a distinct, controlled interval by
+// rewinding the checkpoint, then matched against its own sum and count counters.
+func TestRegionHeartbeatTracerRecordsEachStageIndependently(t *testing.T) {
+	re := require.New(t)
+	h := &regionHeartbeatProcessTracer{}
+	for i, s := range recordingStages() {
+		interval := time.Duration(i+1) * 100 * time.Millisecond // distinct per stage
+		sumBefore := testutil.ToFloat64(s.sum)
+		countBefore := testutil.ToFloat64(s.count)
+		// No real time is spent: rewinding the checkpoint makes the stage measure
+		// exactly `interval`.
+		h.lastCheckTime = time.Now().Add(-interval)
+		s.finish(h)
+		re.InDelta(interval.Seconds(), testutil.ToFloat64(s.sum)-sumBefore, 0.05, s.name)
+		re.Equal(1.0, testutil.ToFloat64(s.count)-countBefore, s.name)
+	}
+}

--- a/pkg/core/metrics_test.go
+++ b/pkg/core/metrics_test.go
@@ -1,4 +1,4 @@
-// Copyright 2024 TiKV Project Authors.
+// Copyright 2026 TiKV Project Authors.
 //
 // Licensed under the Apache License, Version 2.0 (the "License");
 // you may not use this file except in compliance with the License.

--- a/pkg/mcs/resourcemanager/server/metrics.go
+++ b/pkg/mcs/resourcemanager/server/metrics.go
@@ -455,7 +455,12 @@ func (m *counterMetrics) add(consumption *rmpb.Consumption, controllerConfig *Co
 			m.SQLLayerRUMetrics.Add(sqlRU)
 			m.SQLCPUMetrics.Add(consumption.SqlLayerCpuTimeMs)
 		}
-		m.KvCPUMetrics.Add(consumption.TotalCpuTimeMs - consumption.SqlLayerCpuTimeMs)
+		// A well-behaved client keeps TotalCpuTimeMs >= SqlLayerCpuTimeMs, but the
+		// values are reported by the client and not validated here. Guard against a
+		// negative KV CPU time so a malformed report cannot panic prometheus.Counter.Add.
+		if kvCPUTimeMs := consumption.TotalCpuTimeMs - consumption.SqlLayerCpuTimeMs; kvCPUTimeMs > 0 {
+			m.KvCPUMetrics.Add(kvCPUTimeMs)
+		}
 	}
 	// RPC count info.
 	if consumption.KvReadRpcCount > 0 {

--- a/pkg/mcs/scheduling/server/cluster.go
+++ b/pkg/mcs/scheduling/server/cluster.go
@@ -786,6 +786,7 @@ func (c *Cluster) HandleRegionHeartbeat(region *core.RegionInfo) error {
 	if c.persistConfig.GetScheduleConfig().EnableHeartbeatBreakdownMetrics {
 		tracer = core.NewHeartbeatProcessTracer()
 	}
+	defer tracer.Release()
 	var taskRunner, miscRunner, logRunner ratelimit.Runner
 	taskRunner, miscRunner, logRunner = syncRunner, syncRunner, syncRunner
 	if c.persistConfig.GetScheduleConfig().EnableHeartbeatConcurrentRunner {

--- a/server/cluster/cluster_test.go
+++ b/server/cluster/cluster_test.go
@@ -59,7 +59,6 @@ import (
 	"github.com/tikv/pd/pkg/schedule/placement"
 	"github.com/tikv/pd/pkg/schedule/schedulers"
 	"github.com/tikv/pd/pkg/schedule/types"
-	"github.com/tikv/pd/pkg/slice"
 	"github.com/tikv/pd/pkg/statistics"
 	"github.com/tikv/pd/pkg/statistics/utils"
 	"github.com/tikv/pd/pkg/storage"
@@ -1160,10 +1159,6 @@ func TestRegionHeartbeat(t *testing.T) {
 		ctx.Tracer = tracer
 		re.NoError(cluster.processRegionHeartbeat(ctx, overlapRegion))
 		tracer.OnAllStageFinished()
-		re.Condition(func() bool {
-			fields := tracer.LogFields()
-			return slice.AllOf(fields, func(i int) bool { return fields[i].Integer > 0 })
-		}, "should have stats")
 		region = &metapb.Region{}
 		ok, err = storage.LoadRegion(regions[n-1].GetID(), region)
 		re.False(ok)

